

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>clubcpg.CalculateBinCoverage &mdash; CluBCpG 0.1.11 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/graphviz.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> CluBCpG
          

          
          </a>

          
            
            
              <div class="version">
                0.1.11
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">User Guide Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../intro.html">1. Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage.html">2. Using CluBCpG</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../output_data.html">3. Understanding the CluBCpG output data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage_prelim.html">4. Using CluBCpG with PReLIM</a></li>
</ul>
<p class="caption"><span class="caption-text">Developer Reference:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../API.html">API documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">CluBCpG</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>clubcpg.CalculateBinCoverage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for clubcpg.CalculateBinCoverage</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">clubcpg.ParseBam</span> <span class="k">import</span> <span class="n">BamFileReadParser</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="k">import</span> <span class="n">Pool</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">from</span> <span class="nn">pandas.core.indexes.base</span> <span class="k">import</span> <span class="n">InvalidIndexError</span>


<div class="viewcode-block" id="CalculateCompleteBins"><a class="viewcode-back" href="../../API.html#clubcpg.CalculateBinCoverage.CalculateCompleteBins">[docs]</a><span class="k">class</span> <span class="nc">CalculateCompleteBins</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class to calculate the number of reads covering all CpGs</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="CalculateCompleteBins.__init__"><a class="viewcode-back" href="../../API.html#clubcpg.CalculateBinCoverage.CalculateCompleteBins.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">bin_size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">output_directory</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">number_of_processors</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">mbias_read1_5</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mbias_read1_3</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">mbias_read2_5</span><span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">mbias_read2_3</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">no_overlap</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This class is initialized with a path to a bam file and a bin size</span>

<span class="sd">        :param bam_file: One of the BAM files for analysis to be performed</span>
<span class="sd">        :param bin_size: Size of the bins for the analysis, integer</span>
<span class="sd">        :number_of_processors: How many CPUs to use for parallel computation, default=1</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span> <span class="o">=</span> <span class="n">bam_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bin_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">bin_size</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">number_of_processors</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">number_of_processors</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span> <span class="o">=</span> <span class="n">output_directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bins_no_reads</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bins_yes_reads</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">mbias_read1_5</span> <span class="o">=</span> <span class="n">mbias_read1_5</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mbias_read1_3</span> <span class="o">=</span> <span class="n">mbias_read1_3</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mbias_read2_5</span> <span class="o">=</span> <span class="n">mbias_read2_5</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mbias_read2_3</span> <span class="o">=</span> <span class="n">mbias_read2_3</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">no_overlap</span> <span class="o">=</span> <span class="n">no_overlap</span></div>

<div class="viewcode-block" id="CalculateCompleteBins.calculate_bin_coverage"><a class="viewcode-back" href="../../API.html#clubcpg.CalculateBinCoverage.CalculateCompleteBins.calculate_bin_coverage">[docs]</a>    <span class="k">def</span> <span class="nf">calculate_bin_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">bin</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Take a single bin, return a matrix. This is passed to a multiprocessing Pool.</span>

<span class="sd">        :param bin: Bin should be passed as &quot;Chr19_4343343&quot;</span>
<span class="sd">        :return: pd.DataFrame with rows containing NaNs dropped</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Get reads from bam file</span>
        <span class="n">parser</span> <span class="o">=</span> <span class="n">BamFileReadParser</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mbias_read1_5</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mbias_read1_3</span><span class="p">,</span>
                                   <span class="bp">self</span><span class="o">.</span><span class="n">mbias_read2_5</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mbias_read2_3</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">no_overlap</span><span class="p">)</span>
        <span class="c1"># Split bin into parts</span>
        <span class="n">chromosome</span><span class="p">,</span> <span class="n">bin_location</span> <span class="o">=</span> <span class="nb">bin</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="n">bin_location</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">bin_location</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">reads</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_reads</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">bin_location</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">bin_size</span><span class="p">,</span> <span class="n">bin_location</span><span class="p">)</span>
            <span class="n">matrix</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">create_matrix</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="c1"># No reads are within this window, do nothing</span>
            <span class="c1"># logging.info(&quot;No reads found for bin {}&quot;.format(bin))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">bins_no_reads</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Unknown error: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">bin</span><span class="p">))</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="c1"># drop rows of ALL NaN</span>
        <span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">)</span>

        <span class="c1"># convert to data_frame of 1s and 0s, drop rows with NaN</span>
        <span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>

        <span class="c1"># if matrix is empty, attempt to create it with correction before giving up</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">original_matrix</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="c1"># logging.info(&quot;Attempting correction of CpG positions in bin {}&quot;.format(bin))</span>
            <span class="n">reads</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">correct_cpg_positions</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">matrix</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">create_matrix</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>
            <span class="k">except</span> <span class="n">InvalidIndexError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Invalid Index error when creating matrices at bin </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">bin</span><span class="p">))</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
                <span class="k">return</span> <span class="nb">bin</span><span class="p">,</span> <span class="n">original_matrix</span>
            <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Matrix concat error ar bin </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">bin</span><span class="p">))</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

            <span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Correction attempt at bin </span><span class="si">{}</span><span class="s2">: SUCCESS&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">bin</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Correction attempt at bin </span><span class="si">{}</span><span class="s2">: FAILED&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">bin</span><span class="p">))</span>

        <span class="k">return</span> <span class="nb">bin</span><span class="p">,</span> <span class="n">matrix</span></div>

<div class="viewcode-block" id="CalculateCompleteBins.get_chromosome_lengths"><a class="viewcode-back" href="../../API.html#clubcpg.CalculateBinCoverage.CalculateCompleteBins.get_chromosome_lengths">[docs]</a>    <span class="k">def</span> <span class="nf">get_chromosome_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get dictionary containing lengths of the chromosomes. Uses bam file for reference</span>

<span class="sd">        :return: Dictionary of chromosome lengths, ex: {&quot;chrX&quot;: 222222}</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">parser</span> <span class="o">=</span> <span class="n">BamFileReadParser</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">parser</span><span class="o">.</span><span class="n">OpenBamFile</span><span class="o">.</span><span class="n">references</span><span class="p">,</span> <span class="n">parser</span><span class="o">.</span><span class="n">OpenBamFile</span><span class="o">.</span><span class="n">lengths</span><span class="p">))</span></div>

<div class="viewcode-block" id="CalculateCompleteBins.remove_scaffolds"><a class="viewcode-back" href="../../API.html#clubcpg.CalculateBinCoverage.CalculateCompleteBins.remove_scaffolds">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">remove_scaffolds</span><span class="p">(</span><span class="n">chromosome_len_dict</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a dict containing only the standard chromosomes starting with &quot;chr&quot;</span>

<span class="sd">        :param chromosome_len_dict: A dict generated by get_chromosome_lenghts()</span>
<span class="sd">        :return: a dict containing only chromosomes starting with &quot;chr&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">new_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">chromosome_len_dict</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">chromosome_len_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">key</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">):</span>
                <span class="n">new_dict</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">new_dict</span></div>

<div class="viewcode-block" id="CalculateCompleteBins.generate_bins_list"><a class="viewcode-back" href="../../API.html#clubcpg.CalculateBinCoverage.CalculateCompleteBins.generate_bins_list">[docs]</a>    <span class="k">def</span> <span class="nf">generate_bins_list</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome_len_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get a dict of lists of all bins according to desired bin size for all chromosomes in the passed dict</span>

<span class="sd">        :param chromosome_len_dict: A dict of chromosome length sizes from get_chromosome_lenghts, cleaned up by remove_scaffolds() if desired</span>
<span class="sd">        :return: dict with each key being a chromosome. ex: chr1</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_bins</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">chromosome_len_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">bins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bin_size</span><span class="p">,</span> <span class="n">value</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">bin_size</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">bin_size</span><span class="p">))</span>
            <span class="n">bins</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">key</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">bins</span><span class="p">]</span>
            <span class="n">all_bins</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">bins</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">all_bins</span></div>

<div class="viewcode-block" id="CalculateCompleteBins.analyze_bins"><a class="viewcode-back" href="../../API.html#clubcpg.CalculateBinCoverage.CalculateCompleteBins.analyze_bins">[docs]</a>    <span class="k">def</span> <span class="nf">analyze_bins</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">individual_chrom</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Main function in class. Run the Complete analysis on the data</span>

<span class="sd">        :param individual_chrom: Chromosome to analyze: ie &quot;chr7&quot;</span>
<span class="sd">        :return: filename of the generated report</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Track the progress of the multiprocessing and output</span>
        <span class="k">def</span> <span class="nf">track_progress</span><span class="p">(</span><span class="n">job</span><span class="p">,</span> <span class="n">update_interval</span><span class="o">=</span><span class="mi">60</span><span class="p">):</span>
            <span class="k">while</span> <span class="n">job</span><span class="o">.</span><span class="n">_number_left</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Tasks remaining = </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">job</span><span class="o">.</span><span class="n">_number_left</span> <span class="o">*</span> <span class="n">job</span><span class="o">.</span><span class="n">_chunksize</span><span class="p">))</span>
                <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">update_interval</span><span class="p">)</span>

        <span class="c1"># Get and clean dict of chromosome lenghts, convert to list of bins</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Getting Chromosome lengths from bam files...&quot;</span><span class="p">,</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">chromosome_lengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_chromosome_lengths</span><span class="p">()</span>
        <span class="n">chromosome_lengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_scaffolds</span><span class="p">(</span><span class="n">chromosome_lengths</span><span class="p">)</span>

        <span class="c1"># If one chromosome was specified use only that chromosome</span>
        <span class="k">if</span> <span class="n">individual_chrom</span><span class="p">:</span>
            <span class="n">new</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
            <span class="n">new</span><span class="p">[</span><span class="n">individual_chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">chromosome_lengths</span><span class="p">[</span><span class="n">individual_chrom</span><span class="p">]</span>
            <span class="n">chromosome_lengths</span> <span class="o">=</span> <span class="n">new</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Generating bins for the entire genome...&quot;</span><span class="p">,</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">bins_to_analyze</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_bins_list</span><span class="p">(</span><span class="n">chromosome_lengths</span><span class="p">)</span>

        <span class="c1"># Set up for multiprocessing</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Beginning analysis of bins using </span><span class="si">{}</span><span class="s2"> processors.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">number_of_processors</span><span class="p">),</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;This will take awhile.....&quot;</span><span class="p">,</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Loop over bin dict and pool.map them individually</span>
        <span class="n">final_results</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">bins_to_analyze</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Analyzing chromosome </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">),</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">pool</span> <span class="o">=</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">number_of_processors</span><span class="p">)</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map_async</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">calculate_bin_coverage</span><span class="p">,</span> <span class="n">bins_to_analyze</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>

            <span class="n">track_progress</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>

            <span class="c1"># once done, get results</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>

            <span class="n">final_results</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Finished chromosome </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">),</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Analysis complete&quot;</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Complete.&quot;</span><span class="p">,</span> <span class="n">flush</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Write to output file</span>
        <span class="n">output_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_directory</span><span class="p">,</span> <span class="s2">&quot;CompleteBins.</span><span class="si">{}</span><span class="s2">.</span><span class="si">{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">),</span> <span class="n">individual_chrom</span><span class="p">))</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">final_results</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">result</span><span class="p">:</span>
                    <span class="c1"># bin</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="p">)</span>
                    <span class="c1"># num of reads</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="p">)</span>
                    <span class="c1"># num of CpGs</span>
                    <span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Full read coverage analysis complete!&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">output_file</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Anthony Scott, PhD

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>