
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Welcome to MixAnalysis’s documentation! &#8212; MixAnalysis 0.1.8 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="welcome-to-mixanalysis-s-documentation">
<h1>Welcome to MixAnalysis’s documentation!<a class="headerlink" href="#welcome-to-mixanalysis-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
<span class="target" id="module-MixtureAnalysis.ParseBam"></span><dl class="class">
<dt id="MixtureAnalysis.ParseBam.BamFileReadParser">
<em class="property">class </em><code class="descclassname">MixtureAnalysis.ParseBam.</code><code class="descname">BamFileReadParser</code><span class="sig-paren">(</span><em>bamfile</em>, <em>quality_score</em>, <em>read1_5=None</em>, <em>read1_3=None</em>, <em>read2_5=None</em>, <em>read2_3=None</em>, <em>no_overlap=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ParseBam.html#BamFileReadParser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ParseBam.BamFileReadParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Used to simplify the opening and reading from BAM files. BAMs must be coordinate sorted and indexed.</p>
<dl class="method">
<dt id="MixtureAnalysis.ParseBam.BamFileReadParser.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>bamfile</em>, <em>quality_score</em>, <em>read1_5=None</em>, <em>read1_3=None</em>, <em>read2_5=None</em>, <em>read2_3=None</em>, <em>no_overlap=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ParseBam.html#BamFileReadParser.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ParseBam.BamFileReadParser.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Class used to read WGBSeq reads from a BAM file, extract methylation, and convert into data frame</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>bamfile</strong> – Path to bam file location</li>
<li><strong>quality_score</strong> – Only include reads &gt;= this fastq quality</li>
<li><strong>read1_5</strong> – mbias ignore read1 5’</li>
<li><strong>read1_3</strong> – mbias ignore read1 3’</li>
<li><strong>read2_5</strong> – mbias ignore read2 5’</li>
<li><strong>read2_3</strong> – mbias ignore read2 3’</li>
<li><strong>no_overlap</strong> – bool. If overlap exists between two reads, ignore that region from read 2.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ParseBam.BamFileReadParser.create_matrix">
<code class="descname">create_matrix</code><span class="sig-paren">(</span><em>read_cpgs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ParseBam.html#BamFileReadParser.create_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ParseBam.BamFileReadParser.create_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Converted parsed reads into a pandas dataframe.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>read_cpgs</strong> (<em>iterable</em>) – read CpGs generated by self.parse_reads</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">matrix methylated (1) and unmethylated (0) states</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">pd.DataFrame</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ParseBam.BamFileReadParser.fix_read_overlap">
<code class="descname">fix_read_overlap</code><span class="sig-paren">(</span><em>full_reads</em>, <em>read_cpgs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ParseBam.html#BamFileReadParser.fix_read_overlap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ParseBam.BamFileReadParser.fix_read_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes pysam reads and read_cpgs generated during parse reads and removes any
overlap between read1 and read2. If possible it also stitches read1 and read2 together to create
a super read.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>full_reads</strong> – set of reads generated by self.parse_reads()</li>
<li><strong>read_cpgs</strong> – todoo</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A list in the same format as read_cpgs input, but corrected for paired read overlap</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ParseBam.BamFileReadParser.parse_reads">
<code class="descname">parse_reads</code><span class="sig-paren">(</span><em>chromosome: str</em>, <em>start: int</em>, <em>stop: int</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ParseBam.html#BamFileReadParser.parse_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ParseBam.BamFileReadParser.parse_reads" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chromosome</strong> – chromosome as “chr6”</li>
<li><strong>start</strong> – start coordinate</li>
<li><strong>stop</strong> – end coordinate</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">List of reads and their positional tags as assigned by bismark</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Example:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">MixtureAnalysis.ParseBam</span> <span class="k">import</span> <span class="n">BamFileReadParser</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parser</span> <span class="o">=</span> <span class="n">BamFileReadParser</span><span class="p">(</span><span class="s2">&quot;/path/to/data.BAM&quot;</span><span class="p">,</span> <span class="n">quality_score</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_reads</span><span class="p">(</span><span class="s2">&quot;chr7&quot;</span><span class="p">,</span> <span class="mi">10000</span><span class="p">,</span> <span class="mi">101000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">matrix</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">create_matrix</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<span class="target" id="module-MixtureAnalysis.ClusterReads"></span><dl class="class">
<dt id="MixtureAnalysis.ClusterReads.ClusterReads">
<em class="property">class </em><code class="descclassname">MixtureAnalysis.ClusterReads.</code><code class="descname">ClusterReads</code><span class="sig-paren">(</span><em>bam_a: str</em>, <em>bam_b=None</em>, <em>bin_size=100</em>, <em>bins_file=None</em>, <em>output_directory=None</em>, <em>num_processors=1</em>, <em>cluster_member_min=4</em>, <em>read_depth_req=10</em>, <em>remove_noise=True</em>, <em>mbias_read1_5=None</em>, <em>mbias_read1_3=None</em>, <em>mbias_read2_5=None</em>, <em>mbias_read2_3=None</em>, <em>suffix=''</em>, <em>no_overlap=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ClusterReads.html#ClusterReads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ClusterReads.ClusterReads" title="Permalink to this definition">¶</a></dt>
<dd><p>This class is used to take a dataframe or matrix of reads and cluster them</p>
<dl class="method">
<dt id="MixtureAnalysis.ClusterReads.ClusterReads.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>bam_a: str</em>, <em>bam_b=None</em>, <em>bin_size=100</em>, <em>bins_file=None</em>, <em>output_directory=None</em>, <em>num_processors=1</em>, <em>cluster_member_min=4</em>, <em>read_depth_req=10</em>, <em>remove_noise=True</em>, <em>mbias_read1_5=None</em>, <em>mbias_read1_3=None</em>, <em>mbias_read2_5=None</em>, <em>mbias_read2_3=None</em>, <em>suffix=''</em>, <em>no_overlap=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ClusterReads.html#ClusterReads.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ClusterReads.ClusterReads.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ClusterReads.ClusterReads.process_bins">
<code class="descname">process_bins</code><span class="sig-paren">(</span><em>bin</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ClusterReads.html#ClusterReads.process_bins"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ClusterReads.ClusterReads.process_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the main method and should be called using Pool.map It takes one bin location and uses the other helper
functions to get the reads, form the matrix, cluster it with DBSCAN, and output the cluster data as text lines
ready to writing to a file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>bin</strong> – string in this format: “chr19_55555”</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a list of lines representing the cluster data from that bin</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="MixtureAnalysis.ClusterReads.ClusterReadsWithImputation">
<em class="property">class </em><code class="descclassname">MixtureAnalysis.ClusterReads.</code><code class="descname">ClusterReadsWithImputation</code><span class="sig-paren">(</span><em>bam_a: str</em>, <em>bam_b=None</em>, <em>bin_size=100</em>, <em>bins_file=None</em>, <em>output_directory=None</em>, <em>num_processors=1</em>, <em>cluster_member_min=4</em>, <em>read_depth_req=10</em>, <em>remove_noise=True</em>, <em>mbias_read1_5=None</em>, <em>mbias_read1_3=None</em>, <em>mbias_read2_5=None</em>, <em>mbias_read2_3=None</em>, <em>suffix=''</em>, <em>no_overlap=True</em>, <em>models_A=None</em>, <em>models_B=None</em>, <em>chunksize=10000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ClusterReads.html#ClusterReadsWithImputation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ClusterReads.ClusterReadsWithImputation" title="Permalink to this definition">¶</a></dt>
<dd><p>This class is used to perfom the same clustering, but also enabled the ability to perform imputation during clustering.
This inherits from <a class="reference internal" href="#MixtureAnalysis.ClusterReads.ClusterReads" title="MixtureAnalysis.ClusterReads.ClusterReads"><code class="xref py py-class docutils literal notranslate"><span class="pre">ClusterReads</span></code></a></p>
<dl class="method">
<dt id="MixtureAnalysis.ClusterReads.ClusterReadsWithImputation.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>bam_a: str</em>, <em>bam_b=None</em>, <em>bin_size=100</em>, <em>bins_file=None</em>, <em>output_directory=None</em>, <em>num_processors=1</em>, <em>cluster_member_min=4</em>, <em>read_depth_req=10</em>, <em>remove_noise=True</em>, <em>mbias_read1_5=None</em>, <em>mbias_read1_3=None</em>, <em>mbias_read2_5=None</em>, <em>mbias_read2_3=None</em>, <em>suffix=''</em>, <em>no_overlap=True</em>, <em>models_A=None</em>, <em>models_B=None</em>, <em>chunksize=10000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ClusterReads.html#ClusterReadsWithImputation.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ClusterReads.ClusterReadsWithImputation.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

</dd></dl>

<span class="target" id="module-MixtureAnalysis.ConnectToCpGNet"></span><dl class="class">
<dt id="MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet">
<em class="property">class </em><code class="descclassname">MixtureAnalysis.ConnectToCpGNet.</code><code class="descname">ImputeWithCpGNet</code><span class="sig-paren">(</span><em>cpg_density: int</em>, <em>bin: str</em>, <em>bin_size=100</em>, <em>path_to_models=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#ImputeWithCpGNet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet" title="Permalink to this definition">¶</a></dt>
<dd><p>DEPRECATED: Please do not use this class.
Class used to connect parts of MixtureAnalysis to the CpGNet package)</p>
<dl class="method">
<dt id="MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>cpg_density: int</em>, <em>bin: str</em>, <em>bin_size=100</em>, <em>path_to_models=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#ImputeWithCpGNet.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Class used to connect this package with CpGNet’s imputation classes</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>cpg_density {int} – number of CpGs in bin
bin {str} – bin in format “chr1_4444”</dd>
<dt>Keyword Arguments:</dt>
<dd>bin_size {int} – size of bin in base-pairs (default: {100})
path_to_models {[str]} – Path to directory containing models built by TrainModels scripy (default: {None})</dd>
<dt>Raises:</dt>
<dd>AttributeError – Raised if a path to the models directory is not specified or does not exist</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet.get_model">
<code class="descname">get_model</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#ImputeWithCpGNet.get_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet.get_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Called during init to load model from disk into instance object</p>
<dl class="docutils">
<dt>Returns:</dt>
<dd>[type] – CpGNet model</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet.impute_matrix">
<code class="descname">impute_matrix</code><span class="sig-paren">(</span><em>matrix</em>, <em>postprocess=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#ImputeWithCpGNet.impute_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet.impute_matrix" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>matrix</strong> – 2D np array with unknowns as -1</li>
<li><strong>positions</strong> – 1D np array of chromosomal positions (ints)</li>
<li><strong>postprocess</strong> – Imply classes from imputation, if False, returns probabilities instead of predictions</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">A matrix imputed with highly confident predictions, less confident regions are assigned np.nan</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet.postprocess">
<code class="descname">postprocess</code><span class="sig-paren">(</span><em>predicted_matrix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#ImputeWithCpGNet.postprocess"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.ImputeWithCpGNet.postprocess" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes array with predicted values and rounds them to 0 or 1 if threshold is exceeded</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>predicted_matrix {[type]} – matrix generated by imputation</dd>
<dt>Returns:</dt>
<dd>[type] – predicted matrix predictions as 1, 0, or NaN</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="MixtureAnalysis.ConnectToCpGNet.TrainWithCpGNet">
<em class="property">class </em><code class="descclassname">MixtureAnalysis.ConnectToCpGNet.</code><code class="descname">TrainWithCpGNet</code><span class="sig-paren">(</span><em>cpg_density=None</em>, <em>save_path=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#TrainWithCpGNet"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.TrainWithCpGNet" title="Permalink to this definition">¶</a></dt>
<dd><p>Used to train models using CpGnet</p>
<dl class="method">
<dt id="MixtureAnalysis.ConnectToCpGNet.TrainWithCpGNet.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>cpg_density=None</em>, <em>save_path=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#TrainWithCpGNet.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.TrainWithCpGNet.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to train a CpGNet model from input data</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>cpg_density</strong> – Number of CpGs</li>
<li><strong>save_path</strong> – Location of folder to save the resulting model files. One per cpg density</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ConnectToCpGNet.TrainWithCpGNet.save_net">
<code class="descname">save_net</code><span class="sig-paren">(</span><em>model</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#TrainWithCpGNet.save_net"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.TrainWithCpGNet.save_net" title="Permalink to this definition">¶</a></dt>
<dd><p>Save the network to a file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>model</strong> – The trained CpGNet model. Located at CpGNet.model</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Path to the saved model</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.ConnectToCpGNet.TrainWithCpGNet.train_model">
<code class="descname">train_model</code><span class="sig-paren">(</span><em>bins: iter</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/ConnectToCpGNet.html#TrainWithCpGNet.train_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.ConnectToCpGNet.TrainWithCpGNet.train_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Train the CpGNet model on a list of provided bins</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>bins</strong> – iterable containing CpG matrices of 1 (methylated), 0 (unmethylated), and -1 (unknown)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Path to the saved model file</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<span class="target" id="module-MixtureAnalysis.Imputation"></span><dl class="class">
<dt id="MixtureAnalysis.Imputation.Imputation">
<em class="property">class </em><code class="descclassname">MixtureAnalysis.Imputation.</code><code class="descname">Imputation</code><span class="sig-paren">(</span><em>cpg_density: int</em>, <em>bam_file: str</em>, <em>mbias_read1_5=None</em>, <em>mbias_read1_3=None</em>, <em>mbias_read2_5=None</em>, <em>mbias_read2_3=None</em>, <em>processes=-1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/Imputation.html#Imputation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.Imputation.Imputation" title="Permalink to this definition">¶</a></dt>
<dd><p>The class providing convienent APIs to train models and impute from models using CpGNet</p>
<dl class="method">
<dt id="MixtureAnalysis.Imputation.Imputation.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>cpg_density: int</em>, <em>bam_file: str</em>, <em>mbias_read1_5=None</em>, <em>mbias_read1_3=None</em>, <em>mbias_read2_5=None</em>, <em>mbias_read2_3=None</em>, <em>processes=-1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/Imputation.html#Imputation.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.Imputation.Imputation.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>[summary]</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>cpg_density {int} – Number of CpGs this class instance will be used for
bam_file {str} – path to the bam file</dd>
<dt>Keyword Arguments:</dt>
<dd>mbias_read1_5 {[type]} – [description] (default: {None})
mbias_read1_3 {[type]} – [description] (default: {None})
mbias_read2_5 {[type]} – [description] (default: {None})
mbias_read2_3 {[type]} – [description] (default: {None})
processes {int} – number or CPUs to use when parallelization can be utilized, default= All available (default: {-1})</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.Imputation.Imputation.extract_matrices">
<code class="descname">extract_matrices</code><span class="sig-paren">(</span><em>coverage_data_frame: pandas.core.frame.DataFrame</em>, <em>sample_limit: int = None</em>, <em>return_bins=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/Imputation.html#Imputation.extract_matrices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.Imputation.Imputation.extract_matrices" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract CpG matrices from bam file.</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>coverage_data_frame {pd.DataFrame} – Output of CalculateCompleteBins read in as a csv file</dd>
<dt>Keyword Arguments:</dt>
<dd>return_bins {bool} – Return the bin location along with the matrix (default: {False})</dd>
<dt>Returns:</dt>
<dd>[tuple] – Returns tuple of (bin, np.array) if returns_bins = True else returns only np.array</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.Imputation.Imputation.impute_from_model">
<code class="descname">impute_from_model</code><span class="sig-paren">(</span><em>models_folder: str</em>, <em>matrices: iter</em>, <em>postprocess=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/Imputation.html#Imputation.impute_from_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.Imputation.Imputation.impute_from_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator to provide imputed matrices on-the-fly</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>models_folder {str} – Path to directory containing trained CpGNet models
matrices {iter} – An iterable containging n x m matrices with n=cpgs and m=reads</dd>
<dt>Keyword Arguments:</dt>
<dd>postprocess {bool} – Round imputed values to 1s and 0s  (default: {True})</dd>
</dl>
</dd></dl>

<dl class="staticmethod">
<dt id="MixtureAnalysis.Imputation.Imputation.postprocess_predictions">
<em class="property">static </em><code class="descname">postprocess_predictions</code><span class="sig-paren">(</span><em>predicted_matrix</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/Imputation.html#Imputation.postprocess_predictions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.Imputation.Imputation.postprocess_predictions" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes array with predicted values and rounds them to 0 or 1 if threshold is exceeded</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>predicted_matrix {[type]} – matrix generated by imputation</dd>
<dt>Returns:</dt>
<dd>[type] – predicted matrix predictions as 1, 0, or NaN</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="MixtureAnalysis.Imputation.Imputation.train_model">
<code class="descname">train_model</code><span class="sig-paren">(</span><em>output_folder: str</em>, <em>matrices: iter</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/MixtureAnalysis/Imputation.html#Imputation.train_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#MixtureAnalysis.Imputation.Imputation.train_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Train a CpGNet model</p>
<dl class="docutils">
<dt>Arguments:</dt>
<dd>output_folder {str} – Folder to save trained models
matrices {iter} – An iterable of CpGMatrices - ideally obtained through Imputation.extract_matrices()</dd>
<dt>Returns:</dt>
<dd>[keras model] – Returns the trained CpGNet model</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">MixAnalysis</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Anthony Scott.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>